Add `impl Add<Self> for Self`, `impl Mul<Self> for Self` for field mo…

huitseeker opened the pull request:
…dels

Contributes to #50.

/cc @Pratyush

huitseeker committed:
Add `impl Add<Self> for Self`, `impl Mul<Self> for Self` for field models

Contributes to #50.

Pratyush commented on algebra/src/curves/sw6/mod.rs:
```diff
@@ -146,7 +146,7 @@ impl SW6 {
         elt_q3.frobenius_map(3);
         // elt_q3_over_elt = elt^(q^3-1)
-        let elt_q3_over_elt = elt_q3 * &elt_inv;
+        let elt_q3_over_elt = elt_q3 * elt_inv;
```
Was this failing to compile after you added the new impls?

huitseeker replied:
Yes, with:
```rust
error[E0277]: cannot multiply `&&fields::models::fp6_2over3::Fp6<fields::sw6::fq6::Fq6Parameters>` to `fields::models::fp6_2over3::Fp6<fields::sw6::fq6::Fq6Parameters>`
   --> algebra/src/curves/sw6/mod.rs:149:38
    |
149 |         let elt_q3_over_elt = elt_q3 * &elt_inv;
    |                                      ^ no implementation for `fields::models::fp6_2over3::Fp6<fields::sw6::fq6::Fq6Parameters> * &&fields::models::fp6_2over3::Fp6<fields::sw6::fq6::Fq6Parameters>`
    |
    = help: the trait `std::ops::Mul<&&fields::models::fp6_2over3::Fp6<fields::sw6::fq6::Fq6Parameters>>` is not implemented for `fields::models::fp6_2over3::Fp6<fields::sw6::fq6::Fq6Parameters>`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
```

Note that `elt_inv: &Fq6`, so that I think the trait bound is `&Fq6: Mul<&'a $0>`, which would not match in the presence of `impl<P: Fp6Parameters> Mul<Fp6<P>> for Fp6<P>`. IIUC, in its absence (i.e. before this commit), autoref would fire, leading the trait system to solve for `Fq6: Mul<&'a $0>` which matches the (single) existing `impl Mul`.

(Yes, I'm praying for the success of rust-lang issue 44619)

Pratyush reviewed (commented):
Thanks for the PR! It looks good modulo the nit

Pratyush commented:
I probably should have asked this earlier, but it would be great to also have the corresponding `AddAssign` and `MulAssign` impls, maybe via blanket impls (`impl<F: Field> AddAssign<F> for F` where we just invoke the existing AddAssign impl). We can also add these in a followup PR if that's better.

huitseeker commented:
@Pratyush In this case, are you opposed to just using macros, à la [auto_impl](https://github.com/auto-impl-rs/auto_impl)? (I'll need to write something custom, as the derive trigger of `auto_impl` won't be available)

Pratyush commented:
Hmm wouldn’t it be possible to just write a blanket impl that just invokes the existing reference-based impls? Then you only need one impl block

huitseeker commented:
@Pratyush one impl block per instance, right? The trait `AddAssign` is non-local to `Field`, so that I'll get https://gist.github.com/huitseeker/bf2a822fdb06fe20bba29b3a82f0ea9f#file-gistfile1-txt-L2-L9 if I add a blanket impl to `field.rs`

Pratyush commented:
Ahh the orphan rule strikes again. Good point. Let's just do a macro, maybe something like what already exists in `{bigint,fields}/macros.rs`.

huitseeker commented:
@Pratyush Note I will need to remove a lot of references in `algebra/src/fields/models/` in the operands to the calls to `mul_assign`. If unmodified, those `self.c0.mul_assign(&element)` will indeed generate 
```rust
  --> algebra/src/fields/models/fp6_3over2.rs:70:28
   |
70 |         self.c0.mul_assign(&element);
   |                            ^^^^^^^^ no implementation for `fields::models::fp2::Fp2<<P as fields::models::fp6_3over2::Fp6Parameters>::Fp2Params> *= &&fields::models::fp2::Fp2<<P as fields::models::fp6_3over2::Fp6Parameters>::Fp2Params>`
```

.. for precisely the reason we explored [above](https://github.com/scipr-lab/zexe/pull/53#discussion_r366333475). There's [a ton](https://gist.github.com/b379430dfd2e6ffc8aa579ea41711b4e) of such instances, so I want to make sure we're on the same page first.

Pratyush commented:
Oops, just saw your last comment. Yeah, that’s fine, if anything it’s more correct now!

huitseeker committed:
Add implementations of `AddAssign<Self>`, `MulAssign<Self>`, for field models.

Those defer to the ones on `&Self`.  Adjust (de-)references in calls
since callee dispatch has higher precedence than trait impl selection.

huitseeker commented:
@Pratyush Done!

Pratyush commented:
Awesome, thanks!
